### PR TITLE
[SPARK-19724][SQL]create a managed table with an existed default table should throw an exception

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?
This PR is to finish https://github.com/apache/spark/pull/17272

This JIRA is a follow up work after SPARK-19583

As we discussed in that PR

The following DDL for a managed table with an existed default location should throw an exception:

CREATE TABLE ... (PARTITIONED BY ...) AS SELECT ...
CREATE TABLE ... (PARTITIONED BY ...)
Currently there are some situations which are not consist with above logic:

CREATE TABLE ... (PARTITIONED BY ...) succeed with an existed default location
situation: for both hive/datasource(with HiveExternalCatalog/InMemoryCatalog)

CREATE TABLE ... (PARTITIONED BY ...) AS SELECT ...
situation: hive table succeed with an existed default location

This PR is going to make above two situations consist with the logic that it should throw an exception
with an existed default location.
## How was this patch tested?

unit test added

### CODE REVIEW

**Comment by gengliangwang** [comment]
retest this please.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 1]
```diff
@@ -103,6 +103,60 @@ class InMemoryCatalogedDDLSuite extends DDLSuite with SharedSQLContext with Befo
     }
   }
 
+  test("CTAS a managed table with the existing empty directory") {
+    val tableLoc = new File(spark.sessionState.catalog.defaultTablePath(TableIdentifier("tab1")))
+    try {
+      tableLoc.mkdir()
+      withTable("tab1") {
+        sql("CREATE TABLE tab1 USING PARQUET AS SELECT 1, 'a'")
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: Move them to `DDLSuite`? Change the format based on `isUsingHiveMetastore`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 2]
```diff
@@ -298,15 +302,28 @@ class SessionCatalog(
...
     }
 
     requireDbExists(db)
     externalCatalog.createTable(newTableDefinition, ignoreIfExists)
   }
 
+  def validateTableLocation(table: CatalogTable, tableIdentifier: TableIdentifier): Unit = {
+    // SPARK-19724: the default location of a managed table should be non-existent or empty.
+   if (table.tableType == CatalogTableType.MANAGED) {
+     val tableLocation = new Path(defaultTablePath(tableIdentifier))
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: We always check `defaultTablePath `? Is that possible, the table location points to the different location?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 3]
```diff
@@ -298,15 +302,28 @@ class SessionCatalog(
...
     } else {
-      tableDefinition.copy(identifier = TableIdentifier(table, Some(db)))
+      tableDefinition.copy(identifier = tableIdentifier)
     }
 
     requireDbExists(db)
     externalCatalog.createTable(newTableDefinition, ignoreIfExists)
   }
 
+  def validateTableLocation(table: CatalogTable, tableIdentifier: TableIdentifier): Unit = {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: `CatalogTable ` already contains `TableIdentifier `. What is the reason you do not use the one directly?

**Comment by gatorsmile** [comment]
This is a behavior change, we need to make it configurable and also document it in the migration guide.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: I think managed tables are always created on default path. So the check here should be correct.

**Comment by gengliangwang** [comment]
Jenkins, retest this please. 
(Test passed, but jenkins failed to notify github)

**Comment by gengliangwang** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -1152,6 +1152,13 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val ALLOW_NONEMPTY_MANAGED_TABLE_LOCATION =
+    buildConf("spark.sql.allowNonemptyManagedTableLocation")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `spark.sql.allowCreateManagedTableUsingNonemptyLocation`

Also this should be an internal conf

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 5]
```diff
@@ -298,15 +299,32 @@ class SessionCatalog(
...
 
+  def validateTableLocation(table: CatalogTable): Unit = {
+    // SPARK-19724: the default location of a managed table should be non-existent or empty.
+    if (table.tableType == CatalogTableType.MANAGED && !conf.allowNonemptyManagedTableLocation) {
+      val tableLocation =
+        new Path(table.storage.locationUri.getOrElse(defaultTablePath(table.identifier)))
+      val fs = tableLocation.getFileSystem(hadoopConf)
+
+      if (fs.exists(tableLocation) && fs.listStatus(tableLocation).nonEmpty) {
+        throw new AnalysisException(s"Can not create the managed table('${table.identifier}')" +
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: `Can not` -> `Not allowed to`

**Comment by gengliangwang** [comment]
retest this please.

**Pull request closed without merging**